File: backend/app/api/v1/endpoints/artifacts.py
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import FileResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel
from app.core.database import get_db
from app.models.models import Artifact, Task
from app.api.v1.endpoints.auth import get_current_active_user
from pathlib import Path
from loguru import logger

router = APIRouter()

# Pydantic模型
class ArtifactResponse(BaseModel):
    id: int
    task_id: int
    file_name: str
    file_size: int
    download_url: str
    created_at: datetime
    download_count: int
    is_archived: bool
    task_info: dict

class ArtifactListResponse(BaseModel):
    total: int
    artifacts: List[ArtifactResponse]

# API端点
@router.get("/list", response_model=ArtifactListResponse)
async def list_artifacts(
    skip: int = 0,
    limit: int = 20,
    is_archived: Optional[bool] = False,
    db: AsyncSession = Depends(get_db)
):
    """获取产物列表"""
    # 查询总数
    count_query = select(Artifact).where(Artifact.is_archived == is_archived)
    count_result = await db.execute(count_query)
    total = len(count_result.all())

    # 查询产物列表
    query = select(Artifact).where(
        Artifact.is_archived == is_archived
    ).order_by(desc(Artifact.created_at)).offset(skip).limit(limit)

    result = await db.execute(query)
    artifacts = result.scalars().all()

    # 构建响应
    artifact_responses = []
    for artifact in artifacts:
        # 获取关联的任务信息
        task_result = await db.execute(
            select(Task).where(Task.id == artifact.task_id)
        )
        task = task_result.scalar_one_or_none()

        task_info = {}
        if task:
            task_info = {
                "task_uuid": task.task_uuid,
                "unity_branch": task.unity_branch,
                "flutter_branch": task.flutter_branch,
                "status": task.status.value
            }

        artifact_responses.append(ArtifactResponse(
            id=artifact.id,
            task_id=artifact.task_id,
            file_name=artifact.file_name,
            file_size=artifact.file_size,
            download_url=artifact.download_url,
            created_at=artifact.created_at,
            download_count=artifact.download_count,
            is_archived=artifact.is_archived,
            task_info=task_info
        ))

    return ArtifactListResponse(total=total, artifacts=artifact_responses)

@router.get("/{artifact_id}", response_model=ArtifactResponse)
async def get_artifact(
    artifact_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取产物详情"""
    result = await db.execute(
        select(Artifact).where(Artifact.id == artifact_id)
    )
    artifact = result.scalar_one_or_none()

    if not artifact:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Artifact not found"
        )

    # 获取关联的任务信息
    task_result = await db.execute(
        select(Task).where(Task.id == artifact.task_id)
    )
    task = task_result.scalar_one_or_none()

    task_info = {}
    if task:
        task_info = {
            "task_uuid": task.task_uuid,
            "unity_branch": task.unity_branch,
            "flutter_branch": task.flutter_branch,
            "status": task.status.value
        }

    return ArtifactResponse(
        id=artifact.id,
        task_id=artifact.task_id,
        file_name=artifact.file_name,
        file_size=artifact.file_size,
        download_url=artifact.download_url,
        created_at=artifact.created_at,
        download_count=artifact.download_count,
        is_archived=artifact.is_archived,
        task_info=task_info
    )

@router.get("/{artifact_id}/download")
async def download_artifact(
    artifact_id: int,
    db: AsyncSession = Depends(get_db)
):
    """下载产物文件"""
    result = await db.execute(
        select(Artifact).where(Artifact.id == artifact_id)
    )
    artifact = result.scalar_one_or_none()

    if not artifact:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Artifact not found"
        )

    # 检查文件是否存在
    file_path = Path(artifact.file_path)
    if not file_path.exists():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Artifact file not found on disk"
        )

    # 更新下载计数
    artifact.download_count += 1
    artifact.last_downloaded_at = datetime.utcnow()
    await db.commit()

    logger.info(f"Artifact {artifact_id} downloaded")

    return FileResponse(
        path=str(file_path),
        filename=artifact.file_name,
        media_type='application/vnd.android.package-archive'
    )

@router.delete("/{artifact_id}")
async def delete_artifact(
    artifact_id: int,
    current_user = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """删除产物（需要管理员权限）"""
    # 检查权限
    if current_user.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only administrators can delete artifacts"
        )

    result = await db.execute(
        select(Artifact).where(Artifact.id == artifact_id)
    )
    artifact = result.scalar_one_or_none()

    if not artifact:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Artifact not found"
        )

    # 删除文件
    file_path = Path(artifact.file_path)
    if file_path.exists():
        file_path.unlink()

    # 删除数据库记录
    await db.delete(artifact)
    await db.commit()

    logger.info(f"Artifact {artifact_id} deleted by {current_user.username}")
    return {"message": "Artifact deleted successfully"}

@router.post("/{artifact_id}/archive")
async def archive_artifact(
    artifact_id: int,
    current_user = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """归档产物"""
    result = await db.execute(
        select(Artifact).where(Artifact.id == artifact_id)
    )
    artifact = result.scalar_one_or_none()

    if not artifact:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Artifact not found"
        )

    if artifact.is_archived:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Artifact is already archived"
        )

    # 归档产物
    artifact.is_archived = True
    artifact.archived_at = datetime.utcnow()
    await db.commit()

    logger.info(f"Artifact {artifact_id} archived by {current_user.username}")
    return {"message": "Artifact archived successfully"}

File: backend/src/routes/artifacts.ts
import { Router } from 'express';
import { param, query } from 'express-validator';
import path from 'path';
import fs from 'fs/promises';
import { getPrismaClient } from '../services/database';
import { AppError, asyncHandler } from '../middlewares/error';
import { authenticate, authorize, AuthRequest } from '../middlewares/auth';
import { UserRole } from '@prisma/client';
import { logger } from '../utils/logger';

const router = Router();

// Get artifacts list
router.get(
  '/list',
  [
    query('page').optional().isInt({ min: 1 }),
    query('limit').optional().isInt({ min: 1, max: 100 }),
    query('isArchived').optional().isBoolean(),
  ],
  asyncHandler(async (req: any, res: any) => {
    const page = parseInt(req.query.page || '1');
    const limit = parseInt(req.query.limit || '20');
    const isArchived = req.query.isArchived === 'true';

    const prisma = getPrismaClient();

    const where = { isArchived };

    const [artifacts, total] = await Promise.all([
      prisma.artifact.findMany({
        where,
        include: {
          task: {
            select: {
              id: true,
              taskUuid: true,
              unityBranch: true,
              flutterBranch: true,
              status: true,
              initiator: {
                select: {
                  username: true,
                },
              },
            },
          },
        },
        orderBy: {
          createdAt: 'desc',
        },
        skip: (page - 1) * limit,
        take: limit,
      }),
      prisma.artifact.count({ where }),
    ]);

    res.json({
      artifacts,
      pagination: {
        total,
        page,
        limit,
        totalPages: Math.ceil(total / limit),
      },
    });
  })
);

// Get artifact by ID
router.get(
  '/:id',
  param('id').isInt(),
  asyncHandler(async (req: any, res: any) => {
    const artifactId = parseInt(req.params.id);

    const prisma = getPrismaClient();

    const artifact = await prisma.artifact.findUnique({
      where: { id: artifactId },
      include: {
        task: {
          select: {
            id: true,
            taskUuid: true,
            unityBranch: true,
            flutterBranch: true,
            status: true,
            createdAt: true,
            initiator: {
              select: {
                username: true,
              },
            },
          },
        },
      },
    });

    if (!artifact) {
      throw new AppError('Artifact not found', 404);
    }

    res.json({ artifact });
  })
);

// Download artifact
router.get(
  '/:id/download',
  param('id').isInt(),
  asyncHandler(async (req: any, res: any) => {
    const artifactId = parseInt(req.params.id);

    const prisma = getPrismaClient();

    const artifact = await prisma.artifact.findUnique({
      where: { id: artifactId },
    });

    if (!artifact) {
      throw new AppError('Artifact not found', 404);
    }

    // Check if file exists
    try {
      await fs.access(artifact.filePath);
    } catch {
      throw new AppError('Artifact file not found on disk', 404);
    }

    // Update download count
    await prisma.artifact.update({
      where: { id: artifactId },
      data: {
        downloadCount: {
          increment: 1,
        },
        lastDownloadedAt: new Date(),
      },
    });

    logger.info(`Artifact ${artifactId} downloaded`);

    // Send file
    res.download(artifact.filePath, artifact.fileName);
  })
);

// Archive artifact
router.post(
  '/:id/archive',
  authenticate,
  param('id').isInt(),
  asyncHandler(async (req: AuthRequest, res: any) => {
    const artifactId = parseInt(req.params.id);

    const prisma = getPrismaClient();

    const artifact = await prisma.artifact.findUnique({
      where: { id: artifactId },
    });

    if (!artifact) {
      throw new AppError('Artifact not found', 404);
    }

    if (artifact.isArchived) {
      throw new AppError('Artifact is already archived', 400);
    }

    await prisma.artifact.update({
      where: { id: artifactId },
      data: {
        isArchived: true,
        archivedAt: new Date(),
      },
    });

    // Log audit
    await prisma.auditLog.create({
      data: {
        userId: req.user!.id,
        action: 'archive_artifact',
        resourceType: 'artifact',
        resourceId: artifactId.toString(),
        result: 'success',
      },
    });

    logger.info(`Artifact ${artifactId} archived by ${req.user!.username}`);

    res.json({ message: 'Artifact archived successfully' });
  })
);

// Delete artifact (admin only)
router.delete(
  '/:id',
  authenticate,
  authorize(UserRole.ADMIN),
  param('id').isInt(),
  asyncHandler(async (req: AuthRequest, res: any) => {
    const artifactId = parseInt(req.params.id);

    const prisma = getPrismaClient();

    const artifact = await prisma.artifact.findUnique({
      where: { id: artifactId },
    });

    if (!artifact) {
      throw new AppError('Artifact not found', 404);
    }

    // Delete file from disk
    try {
      await fs.unlink(artifact.filePath);
    } catch (error) {
      logger.error(`Failed to delete artifact file: ${artifact.filePath}`, error);
    }

    // Delete from database
    await prisma.artifact.delete({
      where: { id: artifactId },
    });

    // Log audit
    await prisma.auditLog.create({
      data: {
        userId: req.user!.id,
        action: 'delete_artifact',
        resourceType: 'artifact',
        resourceId: artifactId.toString(),
        result: 'success',
      },
    });

    logger.info(`Artifact ${artifactId} deleted by ${req.user!.username}`);

    res.json({ message: 'Artifact deleted successfully' });
  })
);

// Cleanup old artifacts (admin only)
router.post(
  '/cleanup',
  authenticate,
  authorize(UserRole.ADMIN),
  asyncHandler(async (req: AuthRequest, res: any) => {
    const { days = 30, count = 100 } = req.body;

    const prisma = getPrismaClient();

    // Find artifacts to delete
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - days);

    const artifactsToDelete = await prisma.artifact.findMany({
      where: {
        OR: [
          {
            createdAt: {
              lt: cutoffDate,
            },
          },
        ],
      },
      orderBy: {
        createdAt: 'asc',
      },
      skip: count,
    });

    // Delete artifacts
    let deletedCount = 0;
    for (const artifact of artifactsToDelete) {
      try {
        await fs.unlink(artifact.filePath);
      } catch (error) {
        logger.error(`Failed to delete artifact file: ${artifact.filePath}`, error);
      }

      await prisma.artifact.delete({
        where: { id: artifact.id },
      });

      deletedCount++;
    }

    // Log audit
    await prisma.auditLog.create({
      data: {
        userId: req.user!.id,
        action: 'cleanup_artifacts',
        resourceType: 'artifact',
        payload: { days, count, deletedCount },
        result: 'success',
      },
    });

    logger.info(`Artifact cleanup completed by ${req.user!.username}: ${deletedCount} artifacts deleted`);

    res.json({
      message: 'Cleanup completed',
      deletedCount,
    });
  })
);

export default router;